src/Illuminate/Foundation/Application.php

```diff
@@ -530,6 +536,15 @@ protected function markAsRegistered($provider)
 		$this->serviceProviders[] = $provider;
 
 		$this->loadedProviders[$class] = true;
+		
+		$aliases = class_parents($class) + class_implements($class) + array($class);
+		foreach ($aliases as $alias)
+		{
+			if ( ! isset($this->registeredProviders[$alias]))
+			{
+				$this->registeredProviders[$alias] = $provider;
```

Comment:
Unfortunately, I haven't had time to test this.  I know the current change improves things in any case, so I'm sure it could be researched later.